### ISSUE TITLE
HTTP/2 codec synchronization and thread safety for external interface

### ISSUE THREAD

**Issue by Scottmitch** [netty/netty#3196]
A recent issue (https://github.com/netty/netty/pull/3187#discussion_r21184470) has brought to light that care must be taken with the externally available interfaces provided by the codec.  Investigation should be done into the public interfaces to ensure that we account for access from different thread contexts.

**Comment by Scottmitch** [comment]
@nmittler - Do you feel that this should be a candidate for 4.1? It looked like you had things pretty well isolated in gRPC but what is your exposure to this?

**Comment by nmittler** [comment]
@Scottmitch this could be a significant effort. I'm not sure I understand the motivation for this given the nature of the Netty pipeline in 4.0+.

**Comment by Scottmitch** [comment]
@nmittler - IIRC this issue is less about the pipeline and channels and more about the public interfaces we expose and how that relates to the state we keep.  For example I recall making some variables `volatile` in the [DefaultHttp2LocalFlowController](https://github.com/netty/netty/blob/master/codec-http2/src/main/java/io/netty/handler/codec/http2/DefaultHttp2LocalFlowController.java#L47) because we allow them to be manipulated through our public API (separate from the pipeline and channel events).

**Comment by nmittler** [comment]
@Scottmitch Yeah I recall that, although I'm not sure I entirely agree with the rationale. What is the use case for this? Would it be sufficient to document these interfaces as not thread-safe?

**Comment by Scottmitch** [comment]
@nmittler - I think we need some story for this.  Whether it is making thread safe...or documenting that it is not.  I think most of Netty takes the approach that public interfaces (outside of pipeline stuff) are thread safe.

**Comment by nmittler** [comment]
@louiscryan @normanmaurer any thoughts on this?

**Comment by normanmaurer** [comment]
@Scottmitch @nmittler I think not have them thread-safe is fine as long as it is clear. So whatever makes more sense.

**Comment by louiscryan** [comment]
Documentation seems like the right approach unless we want to add significant delay to releasing 4.1.

**Comment by Scottmitch** [comment]
+1 for documenting...and I think we can get rid of volatile variables now :)

**Comment by normanmaurer** [comment]
+1 :)

**Comment by nmittler** [comment]
+1

**Comment by normanmaurer** [comment]
@Scottmitch is this still on your to do list ? I think we need this for 4.1.0.Final
